generative_data_prep/processors/article_tokenizer.py

```diff
@@ -141,6 +147,193 @@ def _update_token_metrics(self, tokenized_sequences: List[TokenizedSequence]):
             self.metrics.completion_tokens += seq.completion_tokens
             self.metrics.padding_tokens += seq.pad_tokens
 
+    def process_jsonl_with_chat_template(self, jsonl: List[dict]) -> List[TokenizedArticle]:
+        """Tokenize a loaded jsonl with the chat template applied and store in a TokenizedArticle object.
+
+        This function processes a list of JSONL (JSON Lines text format) data and converts it into a chat template format.
+        It then tokenizes the chat and returns a list of tokenized articles based on self.BoundaryType.
+        If self.packing_boundary is BoundaryType.PROMPT_COMPLETION_PAIR, then each tokenized article
+        will contain a single tokenized prompt completion pair from the original jsonl.
+        If self.packing_boundary is BoundaryType.JSONL, then each tokenized article will contain
+        the tokenized text from all the prompt completion pairs in the jsonl.
+
+        Args:
+            jsonl (List[dict]): A list of dictionaries where each dictionary represents a JSONL data item.
+            Each dictionary should contain 'prompt' and 'completion' keys.
+            The loaded input jsonl, in the form of {"prompt":"...", "completion":"..."}
+            or [{"prompt":"...", "completion":"..."}, {"prompt":"...", "completion":"..."}, ...]
+
+        Returns:
+            List[dict]: Tokenized articles that represent input jsonl.
+            A list of dictionaries where each dictionary represents a tokenized article.
+        """
+
+        if isinstance(jsonl, dict):
+            jsonl = [jsonl]
+
+        converted_jsonl = []
+        prompts = []
+        completions = []
+
+        # TODO: Add a trick here to get the generation_prompt in tokenizer
+        # This is a temporary solution, and the tokenizer should be modified in the future.
+        # The tokenizer should be able to get the generation_prompt from the input text.
+        # self.tokenizer.chat_template = "{% for message in messages %} {{'<|im_start|>' + message['role'] + '\n' + message['content'] + '<|im_end|>' + '\n'}}{% endfor %}"
+
+        """
+        The mechanism designed here is to remember the original prompt and completion,
+        despite the transformation of the [prompt, completion] pair into a raw string format with additional instruction tokens.
+
+        The following steps are performed:
+        1. Recognizes and preserves the original dialogue data format. (<prompt_placeholder>, <completion_placeholder>)
+        2. Converts the original [prompt, completion] format to the standard conversation [role, content] format.
+            {"prompt": "Hello, how are you?", "completion": "I'm doing great"} ->
+                {"role": "user", "content": "Hello, how are you?"}, {"role": "assistant", "content": "I'm doing great."}
+        """
+        for prompt_completion in jsonl:
+            if self.prompt_keyword not in prompt_completion:
```

Comment:
We actually allow empty prompts in the other version of process_jsonl, sometimes users only want to pretrain or mix pretraining data without any prompts. Lets allow them to not provide a prompt, and default to empty string? 

prompt = prompt_completion[self.prompt_keyword] if self.prompt_keyword in prompt_completion else ""

Reply:
agree, I have revised the code with the latest commit.